airbyte-cdk/bulk/toolkits/load-object-storage/src/main/kotlin/io/airbyte/cdk/load/state/object_storage/ObjectStorageDestinationStateManager.kt

```diff
@@ -60,18 +65,21 @@ class ObjectStorageDestinationState(
             .list(prefix)
             .filter { matcher.match(it.key) != null }
             .toList() // Force the list call to complete before initiating metadata calls
-            .mapNotNull { obj ->
-                val generationId =
-                    client
-                        .getMetadata(obj.key)[destinationConfig.generationIdMetadataKey]
-                        ?.toLongOrNull()
-                        ?: 0L
-                if (generationId < stream.minimumGenerationId) {
-                    Pair(generationId, obj)
-                } else {
-                    null
+            .map { obj ->
+                coroutineScope {
+                    async(Dispatchers.IO) {
+                        Pair(
+                            client
+                                .getMetadata(obj.key)[destinationConfig.generationIdMetadataKey]
+                                ?.toLongOrNull()
+                                ?: 0L,
+                            obj
+                        )
+                    }
                 }
             }
+            .awaitAll()
+            .filter { pair -> pair.first < stream.minimumGenerationId }
```

Comment:
Lets move this filter in the map method, we dont need to create the pair for objects that we are going to discard, inside the map method 
```
 if (genId < stream.minimumGenerationId) genId to obj else null
 ```
 
 and then we could add `filterNotNull()` in the end